### PR TITLE
[base-controller] Consistently apply `Record<string, Json>` constraint on `BaseControllerV2` state objects and types

### LINKED ISSUE #3962: [base-controller] Consistently apply `Record<string, Json>` constraint on BaseControllerV2 state objects and types
There are some key `BaseControllerV2` types where the state type is not constrained with `Record<string, Json>`. This constraint should be applied consistently across the codebase.

### PR DESCRIPTION
**Pull request by MajorLift**
## Motivation

The BaseControllerV2 state type should be constrained with `Record<string, Json>`. Some types need to be fixed to achieve this.

#### Added

- Add and export type `StateConstraint`, which is an alias for `Record<string, Json>`.
  - This type represents the narrowest supertype of the state of all controllers. 
  - Importing this type enables controllers to constrain state objects and types to be Json-serializable without having to add `@metamask/utils` as a dependency.

#### Changed

- **BREAKING:** Narrow the return type of the function `deriveStateFromMetadata<ControllerState>` from `Record<string, Json>` to `Record<keyof ControllerState, Json>`.
- Convert interface `StatePropertyMetadata<T extends Json>` into a type.

#### Fixed

- **BREAKING:** Narrow the generic constraint of the `ControllerState` parameter from `Record<string, unknown>` to `Record<string, Json>` for types `ControllerGetStateAction`, `ControllerStateChangeEvent`, `ControllerActions`, `ControllerEvents`.

#### Removed

- **BREAKING:** Remove the deprecated `subscribe` class field from `BaseControllerV2`.

#### Fixed

- **BREAKING:** Fix `GetRateLimitState`, `RateLimitStateChange` types by replacing `RateLimitedApis` with `RateLimitState<RateLimitedApis>` as the state type passed in as generic arguments to `ControllerGetStateAction` and `ControllerStateChangeEvent`.

## References

- Closes #3962
- Should be released together with #3611 and not before.

## Checklist

- [x] I've updated the test suite for new or updated code as appropriate
- [x] I've updated documentation (JSDoc, Markdown, etc.) for new or updated code as appropriate
- [x] I've highlighted breaking changes using the "BREAKING" category above as appropriate

### CODE REVIEW

File: packages/assets-controllers/src/TokensController.ts [hunk 1]
```diff
@@ -111,10 +110,11 @@ export type TokensControllerActions =
   | TokensControllerGetStateAction
   | TokensControllerAddDetectedTokensAction;
 
-export type TokensControllerGetStateAction = ControllerGetStateAction<
-  typeof controllerName,
-  TokensState
->;
+// TODO: Once `TokensController` is upgraded to V2, rewrite this type using the `ControllerGetStateAction` type, which constrains `TokensState` as `Record<string, Json>`.
```
Comment (mcmire) [inline packages/assets-controllers/src/TokensController.ts:113]: 👍🏻

File: packages/rate-limit-controller/CHANGELOG.md [hunk 2]
```diff
@@ -7,6 +7,10 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ## [Unreleased]
 
+### Fixed
+
+- **BREAKING:** Fix `GetRateLimitState`, `RateLimitStateChange` types by replacing `RateLimitedApis` with `RateLimitState<RateLimtiedApis>` as the state type passed in as generic arguments to `ControllerGetStateAction` and `ControllerStateChangeEvent` ([#3949](https://github.com/MetaMask/core/pull/3949))
```
Comment (mcmire) [inline packages/rate-limit-controller/CHANGELOG.md:12]: Nit: Typo

``` suggestion
- **BREAKING:** Fix `GetRateLimitState`, `RateLimitStateChange` types by replacing `RateLimitedApis` with `RateLimitState<RateLimitedApis>` as the state type passed in as generic arguments to `ControllerGetStateAction` and `ControllerStateChangeEvent` ([#3949](https://github.com/MetaMask/core/pull/3949))
- ```

**Review by mcmire — COMMENTED** [review]
Makes sense! I believe there are still some changes that the Confirmation Systems team wants to get released for TransactionController before we can merge this, however. Maybe we should mark this as `DO-NOT-MERGE` for now?

**Review by Mrtenz — APPROVED** [review]

**Comment by MajorLift** [comment]
Removing `DO-NOT-MERGE` label due to core v120 release.

**Pull request merged**